Add repeated mapping to the dynamic task guide (#2117)
astronomer/docs: learn/dynamic-tasks.md

- Map index 1: `1202` (2+1000+200)
- Map index 2: `2300` (1000+1000+300)

> ## Repeated mapping

Reviewer: For my own understanding: is this the same thing as recursive mapping?

Author: Hmm, I think with dynamic task mapping what I'd call recursive mapping is if each mapped task instance would be the base of an entire set of more mapped tasks. so 3 instances, next task maps 3 = 9 and so forth and we would be using the same task, which keeps using its own mapped version as an input until it hits a base case. Which I am pretty sure is currently not possible but would be very interesting (and potentially system crashing 😅 ).

But yeah that is a great question, I think if that exponential pattern becomes possible recursive mapping would be a possible term for it 🤔 . My programming terminology is a little too blurry here to be sure if that would be ok, since usually recursion refers to a function calling itself, so maybe it depends on how the code is resolved under the hood, whether we could truly call that recursion.

Btw I stole the term repeated mapping from the Airflow docs :)

> It is possible to dynamically map an Airflow task over the output of another dynamically mapped task. This results in one mapped task instance for every mapped task instance of the upstream task.

Reviewer: ```suggestion
You can dynamically map an Airflow task over the output of another dynamically mapped task. This results in one mapped task instance for every mapped task instance of the upstream task.
```

    defaultValue="taskflow"
    groupId="repeated-mapping"
    values={[
        {label: 'TaskFlow API', value: 'taskflow'},
        {label: 'Traditional syntax', value: 'traditional'},
    ]}>

<TabItem value="taskflow">

```python
@task
def multiply_by_2(num):
    return num * 2

@task
def add_10(num):
    return num + 10

@task
def multiply_by_100(num):
    return num * 100

multiply_by_100.expand(
    num=add_10.expand(
        num=multiply_by_2.expand(
            num=[1, 2, 3]
        )
    )
)
```
</TabItem>
<TabItem value="traditional">

```python
def multiply_by_2_func(num):
    return [num * 2]

def add_10_func(num):
    return [num + 10]

def multiply_by_100_func(num):
    return num * 100

multiply_by_2 = PythonOperator.partial(
    task_id="multiply_by_2",
    python_callable=multiply_by_2_func
).expand(op_args=[[1], [2], [3]])

add_10 = PythonOperator.partial(
    task_id="add_10",
    python_callable=add_10_func
).expand(op_args=multiply_by_2.output)

multiply_by_100 = PythonOperator.partial(
    task_id="multiply_by_100",
    python_callable=multiply_by_100_func
).expand(op_args=add_10.output)

multiply_by_2 >> add_10 >> multiply_by_100
```
</TabItem>
</Tabs>

> In the example above, the `multiply_by_2` task is dynamically mapped over a list of 3 elements (`[1, 2, 3]`). The task has 3 mapped task instances returning the following values:

Reviewer: ```suggestion
In the example above, the `multiply_by_2` task is dynamically mapped over a list of three elements (`[1, 2, 3]`). The task has three mapped task instances containing the following values:
```

Reviewer: Feel free to change the word back if my understanding is wrong.

Author: That is a good question. I would say the value is returned by the function that the task uses, which pushes it to XCom with the return_value key, which is the can then be retrieved from the mapped task instance by calling xcom_pull, so I think saying that it is contained by the task instance works. :) 

(looking at the source code XCom are evaluated in a lazy way that I don't fully understand yet, so this might not be entirely true, and maybe some of the values are only created/returned if they are actually accessed, but I think  contain works for now :) )

- Map index 0: `2` (1*2)
- Map index 1: `4` (2*2)
- Map index 2: `6` (3*2)

> The `add_10` task is dynamically mapped over the output of the `multiply_by_2` task. It has 3 mapped task instances (one for each mapped instance of the previous task) returning the values:

Reviewer: ```suggestion
The `add_10` task is dynamically mapped over the output of the `multiply_by_2` task. It has 3 mapped task instances (one for each mapped instance of the previous task) which contain the following values:
```

> The `add_10` task is dynamically mapped over the output of the `multiply_by_2` task. It has 3 dynamically mapped task instances (one for each mapped instance of the previous task) returning the values:

Reviewer 2: ```suggestion
The `add_10` task is dynamically mapped over the output of the `multiply_by_2` task. It has 3 mapped task instances (one for each mapped instance of the previous task) returning the values:
```

- Map index 0: `12` (2+10)
- Map index 1: `14` (4+10)
- Map index 2: `16` (6+10)

> The `multiply_by_100` task is dynamically mapped over the output of the `add_10` task. It also has 3 dynamically mapped task instances returning the values:

Reviewer 2: ```suggestion
The `multiply_by_100` task is dynamically mapped over the output of the `add_10` task. It also has 3 mapped task instances returning the values:
```
